garage/envs/mujoco/sawyer/reacher_env.py

```diff
@@ -0,0 +1,144 @@
+"""Reacher environment for the sawyer robot."""
+
+from gym.spaces import Box
+import numpy as np
+
+from garage.core import Serializable
+from garage.envs import Step
+from garage.envs.mujoco import MujocoEnv
+from garage.misc import overrides
+
+
+class ReachEnv(MujocoEnv, Serializable):
+    """Reacher Environment."""
+
+    FILE = "reach.xml"
+
+    def __init__(self,
+                 initial_goal,
+                 initial_qpos,
+                 distance_threshold=0.05,
+                 target_range=0.15,
+                 sparse_reward=True,
+                 *args,
+                 **kwargs):
+        """
+        Reacher Environment.
+
+        :param initial_goal: initial position to reach.
+        :param initial_qpos: initial qpos for each joint.
+        :param distance_threshold: distance threhold to define reached.
+        :param target_range: delta range the goal is randomized
+        :param sparse_reward: whether using sparse reward
+        :param args
+        :param kwargs
+        """
+        Serializable.quick_init(self, locals())
+        self._initial_goal = initial_goal
+        self._distance_threshold = distance_threshold
+        self._target_range = target_range
+        self._sparse_reward = sparse_reward
+
+        self._goal = self._initial_goal
+        super(ReachEnv, self).__init__(*args, **kwargs)
+        self._env_setup(initial_qpos)
+
+    @overrides
+    def step(self, action):
+        """
+        Perform one step with action.
+
+        :param action: the action to be performed
+        :return: next_obs, reward, done, info
+        """
+        self.forward_dynamics(action)
+
+        obs = self.get_current_obs()
+        next_obs = obs['observation']
+        achieved_goal = obs['achieved_goal']
+        goal = obs['desired_goal']
+        reward = self._compute_reward(achieved_goal, goal)
+        done = (self._goal_distance(achieved_goal, goal) <
+                self._distance_threshold)
+
+        return Step(next_obs, reward, done)
+
+    def _compute_reward(self, achieved_goal, goal):
+        # Compute distance between goal and the achieved goal.
+        d = self._goal_distance(achieved_goal, goal)
+        if self._sparse_reward:
+            return -(d > self._distance_threshold).astype(np.float32)
```

Comment:
Can you add a completion bonus? We find it's usually necessary to guarantee stability.

I usually set the completion bonus large enough to make the AverageReturn guaranteed positive for an agent which has masted the environment (e.g. it wipes out the negative path rewards accrued while moving to the goal).

Reply:
I think for sparse rewards, this implementation is good.
For dense rewards, it would actually be nice to compute rewards in [0, 1] because most algorithms need this normalized reward. DM Control Suite does this using [sigmoid functions](https://github.com/deepmind/dm_control/blob/master/dm_control/utils/rewards.py). The env normalizer can to do that by batch-normalizing the rewards which results in a constant rescaling which impacts the training.
I agree with Ryan on the completion bonus - it should be something like the absolute sum of rewards that a typical successful policy can accrue until it hits the target.